### PR TITLE
Raise an error if an empty `UserPrincipal` is passed to `ClientConfigWithUser`

### PR DESCRIPTION
**Pull request by SamLR**
**What changed?**

Raise an error if an empty `UserPrincipal` is passed to `ClientConfigWithUser`. 

**Why was this change made?**

Most callers of `ClientConfigWithUser` check that the user principal they're about to pass isn't `nil` but this doesn't protect against the case when both the `Token` and `ID` are empty strings. If this happens then any calls to the Kubernetes API are made with the application's service account rather than the (incorrectly configured) user.

This is (unlikely) to pose a significant risk as other layers ensure that a principal is set correctly but this provides strength in depth against potential miss-configuration or future extensions.

**How was this change implemented?**

A check for empty strings.

**How did you validate the change?**

Extended the unit tests and running locally

### CODE REVIEW

File: core/clustersmngr/clustersmngr.go [hunk 1]
```diff
@@ -90,7 +90,9 @@ func ClientConfigWithUser(user *auth.UserPrincipal) ClusterClientConfigFunc {
 			}).DialContext,
 		}
 
-		if user.Token != "" {
+		if user.Token == "" && user.ID == "" {
```
Comment (bigkevmcd) [inline core/clustersmngr/clustersmngr.go:93]: Maybe we could move this to `.Validate()` on the Principal?

Comment (SamLR) [reply core/clustersmngr/clustersmngr.go:93]: We could but as far as I'm aware nothing else uses this and it feels like it makes more sense to keep it inline where it's clear what's being checked for and why.

**Comment by SamLR** [comment]
@bigkevmcd are you happy for me to merge this?

**Comment by bigkevmcd** [comment]
> @bigkevmcd are you happy for me to merge this?

Sure, I would prefer `.Valid()` because it makes it more clear what it's doing ("Does it only need to check these fields in this case?"), saves someone else adding the same checking again, and makes it clear where to add other validation items, but I don't feel that this should be a blocker.

**Comment by SamLR** [comment]
Thinking about it I was just being lazy (and this wasn't actually much of a change) so I've added a `valid` method and some tests

**Review by bigkevmcd — APPROVED** [review]
Thank you!

**Review by ozamosi — APPROVED** [review]

**Pull request merged**